src/detect.rs

```diff
@@ -18,3 +23,221 @@ pub enum DetectOutcome {
+    ///
+    /// let temp_dir = tempfile::tempdir().unwrap();
+    /// let context = DetectContextBuilder::new("heroku-20")
+    ///     .buildpack_toml_from_string(include_str!("../examples/example-02-ruby-sample/buildpack.toml"))
+    ///     .from_temp_dir(&temp_dir)
+    ///     .build::<std::io::Error>()
+    ///     .unwrap();
+    ///
+    /// detect(context);
+    ///```
+    pub fn from_temp_dir(mut self, temp_dir: &tempfile::TempDir) -> DetectContextBuilder<P, BM> {
+        let app_dir = temp_dir.path().join("app");
+        let platform_dir = temp_dir.path().join("platform");
+        let buildpack_dir = temp_dir.path().join("buildpack");
+
+        for dir in [&app_dir, &platform_dir, &buildpack_dir] {
+            std::fs::create_dir_all(dir).unwrap();
+        }
+
+        if self.app_dir.is_none() {
+            self.app_dir = Some(app_dir);
+        }
+
+        if self.buildpack_dir.is_none() {
+            self.buildpack_dir = Some(buildpack_dir);
+        }
+
+        if self.platform.is_none() {
+            self.platform_dir = Some(platform_dir);
+        }
+        self
+    }
+
+    pub fn buildpack_toml_from_string(
+        mut self,
+        body: impl AsRef<str>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml_string = Some(String::from(body.as_ref()));
+        self
+    }
+
+    pub fn buildpack_toml_from_file(
+        mut self,
+        path: impl AsRef<Path>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml_path = Some(path.as_ref().to_path_buf());
+        self
+    }
+
+    pub fn buildpack_toml(
+        mut self,
+        buildpack_toml: BuildpackToml<BM>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(buildpack_toml);
+        self
+    }
+
+    pub fn platform(mut self, platform: P) -> DetectContextBuilder<P, BM> {
+        self.platform = Some(platform);
+        self
+    }
+
+    pub fn app_dir(mut self, app_dir: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
+        self.app_dir = Some(app_dir.as_ref().to_path_buf());
+        self
+    }
+
+    pub fn buildpack_dir(mut self, buildpack_dir: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
+        self.buildpack_dir = Some(buildpack_dir.as_ref().to_path_buf());
+        self
+    }
+
+    pub fn platform_from_path(
+        mut self,
+        platform_dir: impl AsRef<Path>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.platform_dir = Some(platform_dir.as_ref().to_path_buf());
+        self
+    }
+
+    pub fn build<E: Display + Debug>(self) -> crate::Result<DetectContext<P, BM>, E> {
+        match vec![self.platform.is_some(), self.platform_dir.is_some()]
+            .iter()
+            .filter(|&&x| x)
+            .count()
+        {
+            0 => {
+                return Err(crate::error::Error::Other(
+                    "Platform is missing, call platform() or platform_from_path()".to_string(),
```

Comment:
If we want to use this to replace code internally, I'm not sure this is a great error message to return. This feels very specific to testing.

Reply:
It's a very internal error message. The flexibility that using a builder buys you means that it can also fail in ways not previously possible. When it fails I want to give the implementer more info on how to fix it. I'm not sure how to better guard to make sure that end buildpack authors don't see this (unless in testing).

Reply:
I updated the error handling, please take another look.